packages/eslint-plugin/src/rules/no-void-expression.ts

```diff
@@ -0,0 +1,284 @@
+    },
+    messages: {
+      invalidVoidExpr:
+        'Placing a void expression inside another expression is forbidden. ' +
+        "Move it to it's own statement instead.",
+      invalidVoidExprWrapVoid:
+        'Void expressions used inside another expression ' +
+        'must be marked explicitly with the `void` operator.',
+      invalidVoidExprArrow:
+        'Returning a void expression from an arrow function shorthand is forbidden. ' +
+        'Please add braces to the arrow function.',
+      invalidVoidExprArrowWrapVoid:
+        'Void expressions returned from an arrow function shorthand ' +
+        'must be marked explicitly with the `void` operator.',
+      invalidVoidExprReturn:
+        'Returning a void expression from a function is forbidden.' +
+        'Please move it before the `return` statement.',
+      invalidVoidExprReturnLast:
+        'Returning a void expression from a function is forbidden.' +
+        'Please remove the `return` statement.',
+      invalidVoidExprReturnWrapVoid:
+        'Void expressions returned from a function ' +
+        'must be marked explicitly with the `void` operator.',
+      voidExprWrapVoid: 'Mark with an explicit `void` operator',
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          ignoreArrowShorthand: { type: 'boolean' },
+          ignoreVoidOperator: { type: 'boolean' },
+        },
+        additionalProperties: false,
+      },
+    ],
+    type: 'problem',
+    fixable: 'code',
+  },
+  defaultOptions: [{}],
+
+  create(context, [options]) {
+    return {
+      'AwaitExpression, CallExpression, TaggedTemplateExpression'(
+        node:
+          | TSESTree.AwaitExpression
+          | TSESTree.CallExpression
+          | TSESTree.TaggedTemplateExpression,
+      ): void {
+        const parserServices = util.getParserServices(context);
+        const checker = parserServices.program.getTypeChecker();
+        const tsNode = parserServices.esTreeNodeToTSNodeMap.get(node);
+        const type = util.getConstrainedTypeAtLocation(checker, tsNode);
+        if (!tsutils.isTypeFlagSet(type, ts.TypeFlags.VoidLike)) {
+          // not a void expression
+          return;
+        }
+
+        const invalidAncestor = findInvalidAncestor(node);
+        if (invalidAncestor == null) {
+          // void expression is in valid position
+          return;
+        }
+
+        const sourceCode = context.getSourceCode();
+        const wrapVoidFix = (fixer: TSESLint.RuleFixer): TSESLint.RuleFix => {
+          const nodeText = sourceCode.getText(node);
+          const newNodeText = `void ${nodeText}`;
+          return fixer.replaceText(node, newNodeText);
+        };
+
+        // handle arrow function shorthand
+        if (invalidAncestor.type === AST_NODE_TYPES.ArrowFunctionExpression) {
+          // handle wrapping with `void`
+          if (options.ignoreVoidOperator) {
+            return context.report({
```

Comment:
The problem with `return void` is that nobody writes code like that and it might be confusing unless it's enforced in the whole repo. I could change it to `context.report(...);` and then `return;` but I think it's ugly. I added the comment for this reason, but I could change it to the latter style. I think both are explicit enough.

Reply:
that is fair, just thought it was ironic and wanted to ask.  

I do wonder if an option that allows returning `void` expressions inside a function that explicitly has a return annotation of `void` to be allowed since I do think this is a reasonable use case.

Reply:
Actually... Why not enable this rule in it's own implementation only. I added eslint directive at the top and it works nicely. Look now.

Reply:
Also the `eslint-comments/no-use` is super annoying. It should only report when I'm using a directive to disable something. Not when I'm enabling an extra rule for a particular file. cc @bradzacher

Reply:
Feel free to reconfigure it in the root config!
https://mysticatea.github.io/eslint-plugin-eslint-comments/rules/no-use.html#options
`{allow:["eslint-enable"]}`

Reply:
I do want to mention that the sorting of `index.ts` and submitting the new rule may get complications independent of each other so trying to submit both in the same PR may slow things down.  

Personally I would remove the reordering of the rules index from this request and submit a separate PR that does it to prevent merge conflicts, certainly from a reviewing perspective it's less noisy.  Will let you do your thing just speaking from experience.

Reply:
I removed that comment in `src/rules/index.ts` and enforced ordering in the main config via an override:

```js
    // files which list all the things
    {
      files: ['packages/eslint-plugin/src/rules/index.ts'],
      rules: {
        // enforce alphabetical ordering
        'sort-keys': 'error',
        'import/order': ['error', { alphabetize: { order: 'asc' } }],
      },
    },
```

> Personally I would remove the reordering of the rules index from this request and submit a separate PR that does it to prevent merge conflicts

There weren't that many changes. It shouldn't cause any conflicts. I always sort that file when submitting a new rule and nothing bad happens.